raw-ipa-lib/src/user.rs

```diff
@@ -68,18 +69,46 @@ impl User {
         Ok(())
     }
 
-    fn point_from_matchkey(mk: &[u8; 32]) -> RistrettoPoint {
-        // Note that ristretto wants 64 bytes of input; also we don't know if the input is uniform.
-        // TODO: Consider salting this input somehow (with the provider, perhaps).
-        //       The caveat being that anything we do needs to be standardized.
-        RistrettoPoint::hash_from_bytes::<Sha512>(&mk[..])
+    /// This part ultimately needs to be standardized.
+    /// This should do for now in terms of getting a good input.
+    fn point_from_matchkey(provider: &str, mk: &[u8]) -> RistrettoPoint {
+        let mut input = Vec::with_capacity(2 + provider.len() + mk.len());
+        input.push(u8::try_from(provider.len()).unwrap());
+        input.extend_from_slice(provider.as_bytes());
+        input.push(u8::try_from(mk.len()).unwrap());
+        input.extend_from_slice(mk);
+        RistrettoPoint::hash_from_bytes::<Sha512>(&input)
     }
 
-    pub fn set_matchkey(&mut self, provider: &str, mk: &[u8; 32]) {
-        let m = Self::point_from_matchkey(mk);
+    pub fn set_matchkey(&mut self, provider: impl AsRef<str>, mk: impl AsRef<str>) {
+        let m = Self::point_from_matchkey(provider.as_ref(), mk.as_ref().as_bytes());
```

Comment:
This is *slightly* more usable.  `impl Foo` means that the function is generic and will accept anything that implements the type.  In this case, it means that you can pass an owned `String` or `&str` because both implement `AsRef<str>`.  But it also means that you can accept other types (like my new `StringN`) that implement that trait.

Reply:
Great, makes sense. Thanks for educating me!